### PR TITLE
WIP: ENH: fft: support Array API, see #26 [Design 4]

### PR DESCRIPTION
**Pull request by lucascolley**
### Reference issue
see #26 

### What is this working to implement?

`scipy.fft` currently works on the assumption that its inputs are NumPy arrays. Passing in non-NumPy arrays can cause errors, e.g. (where `scipy` is imported as `sp` and `cupy` is imported as `cp`)
```
>>> sp.fft.fft(cp.exp(2j * cp.pi * cp.arange(8) / 8))
[...]
KeyError: 'ALIGNED is not defined for cupy.ndarray.flags'
```
In order to make `scipy.fft` conform to the Array API standard, this solution aims to deal with NumPy arrays and non-NumPy arrays separately, and for non-NumPy arrays which already implement `fft`, to use their implementation. This is visualised for `scipy.fft.fft` in this diagram:
image

This PR proposes to rename the current files which contain the methods by adding the suffix `_np`, replacing them with files of the original name which handle all array types. These new files will import the `_np` files and use the current methods just in the case of NumPy arrays.

### What is the current progress?

The commits that come with this PR have created the new files as per the above paragraph and have attempted to change all of the array API standard functions to support the array API.

**What works?** `python dev.py test -s fft` has all tests pass.

**What is broken?** `python dev.py test -s fft -b cupy` has these two tests fail:
image

**What may be working but hasn't been tested?** The handling of non-NumPy arrays which do not implement `fft`. The design in this case is to (attempt to) convert to a NumPy array, use the `_pocketfft` subpackage, then convert back to the original type.

### Point of interest

In order to handle the case of NumPy arrays separately I am using `if isinstance(x, np.ndarray)` frequently. I do not know whether this is the right way to handle this case, or even whether this case needs to be handled first in every function. Since `np` implements `fft`, I have not checked for this case in `fftfreq` and the other array API standard functions found in `_helper.py`. The rationale for checking this case in the other array API standard functions (those found in `_basic.py`) is that there are keyword arguments included in the SciPy API which are not in the array API. With checking for this case, these arguments can be provided along with NumPy arrays but not with arrays from any other library.

### CODE REVIEW

**Comment by tupui** [comment]
Thanks @lucascolley

I updated my `main` so you can rebase now.

For the failing tests, you might want to add the skipping decorator on these.

The reason it's failing is that in some tests we are checking that passing "array like", here lists of lists of floats, is also working. I suggest you just keep a few of these tests and convert most of them to use real array objects.

File: scipy/fft/_basic.py [hunk 1]
```diff
@@ -1,1629 +1,358 @@
...
+    For non-numpy arrays, this implements the Array API specification of fft.
+    For numpy arrays, see the documentation in _basic_np.py.
+    Note that if arguments outside of those in the Array API specification
+    are provided with a non-numpy array, an exception is raised.
+    """
+    if isinstance(x, np.ndarray):
+        return npbasic.fft(x, n=n, axis=axis, norm=norm, overwrite_x=overwrite_x,
+                           workers=workers, plan=plan)
+    if overwrite_x is not False:
+        Exception
```
Comment (izaid) [inline scipy/fft/_basic.py:19]: What we want to do here is raise an error via `raise SomeError('...some comment')`. I think ValueError is the appropriate one, there should be an example somewhere else in SciPy of a similar error we can use.

File: scipy/fft/_basic.py [hunk 2]
```diff
@@ -1,1629 +1,358 @@
...
+    For non-numpy arrays, this implements the Array API specification of fftn.
+    For numpy arrays, see the documentation in _basic_np.py.
+    Note that if arguments outside of those in the Array API specification
+    are provided with a non-numpy array, an exception is raised.
+    """
+    if isinstance(x, np.ndarray):
+        return npbasic.fftn(x, s=s, axes=axes, norm=norm, overwrite_x=overwrite_x,
+                            workers=workers, plan=plan)
+    if overwrite_x is not False:
+        Exception
```
Comment (izaid) [inline scipy/fft/_basic.py:95]: Same comment on the error as for fft.

File: scipy/fft/_basic.py [hunk 3]
```diff
@@ -1,1629 +1,358 @@
...
-    If the input `a` contains an imaginary part, it is silently discarded.
+    For non-numpy arrays, this implements the Array API specification of ifft.
+    For numpy arrays, see the documentation in _basic_np.py.
+    Note that if arguments outside of those in the Array API specification
+    are provided with a non-numpy array, an exception is raised.
+    """
+    if isinstance(x, np.ndarray):
+        return npbasic.ifft(x, n=n, axis=axis, norm=norm, overwrite_x=overwrite_x,
+                            workers=workers, plan=plan)
+    if overwrite_x is not False:
```
Comment (izaid) [inline scipy/fft/_basic.py:43]: Same comment on the error as for FFT>

**Comment by lucascolley** [comment]
I'm closing this PR now to shift discussion to the new PR on my fork (https://github.com/lucascolley/scipy/pull/2) since we seem set on this design option now and all comments have been noted. I intend to use this new PR as a draft for the full PR until it is ready to be opened on the main repository. Issue #26 remains open for more general discussion, but I expect most updates and discussion to occur on the new PR until it is time to open on the main repo. Cheers!

**Pull request closed without merging**